lib/spack/spack/cmd/env.py

```diff
@@ -613,17 +620,103 @@ def env_depfile_setup_parser(subparser):
         choices=("make",),
         help="specify the depfile type. Currently only make is supported.",
     )
+    subparser.add_argument(
+        metavar="specs",
+        dest="specs",
+        nargs=argparse.REMAINDER,
+        default=None,
+        help="generate a depfile only for matching specs in the environment",
+    )
+
+
+class SpecNode(object):
+    def __init__(self, spec, depth):
+        self.spec = spec
+        self.depth = depth
+
+    def key(self):
+        return self.spec.dag_hash()
+
+
+class UniqueNodesQueue(object):
+    def __init__(self, init=[]):
+        self.seen = set()
+        self.queue = []
+        for item in init:
+            self.push(item)
+
+    def push(self, item):
+        key = item.key()
+        if key in self.seen:
+            return
+        self.queue.append(item)
+        self.seen.add(key)
+
+    def empty(self):
+        return len(self.queue) == 0
+
+    def pop(self):
+        return self.queue.pop()
+
+
+class MakeTargetVisitor(object):
+    """Keep track of hash -> [hashes of prereqs]"""
+
+    def __init__(self, target):
+        # array of (parent hash, human readable spec string, children hashes)
+        self.items = []
+        self.target = target
+
+    def visit(self, node):
+        self.items.append(
+            (
+                node.spec.dag_hash(),
+                node.spec.format("{name}{@version}{%compiler}{variants}{arch=architecture}"),
+                " ".join([self.target(dep.dag_hash()) for dep in self.neighbors(node)]),
+            )
+        )
+
+
+class FullDagNeighborsMixin(object):
+    """Visit each node in the DAG once"""
+
+    def neighbors(self, node):
+        return node.spec.dependencies(deptype="all")
+
+
+class MinimalDagNeighborsMixin(object):
+    """Visit the sub-DAG where non-direct build deps are pruned"""
+
+    def neighbors(self, node):
+        # At depth zero we return build / test deps too
+        if node.depth == 0:
+            return node.spec.dependencies(deptype="all")
+        # Otherwise we just return link/run type deps
+        return node.spec.dependencies(deptype=("link", "run"))
+
+
+class FullMakeTargetVisitor(MakeTargetVisitor, FullDagNeighborsMixin):
+    pass
+
+
+class ShallowMakeTargetVisitor(MakeTargetVisitor, MinimalDagNeighborsMixin):
+    pass
+
+
+def traverse_breadth_first(visitor, specs=[]):
```

Comment:
`Spec.traverse(order='pre')` should match this in all the desired ways and might avoid some corner cases like:

```
1____
| \  \
2  3  4
|  |  |
|  5  6
|_/__/
|
7
```

i.e., we want something more-sophisticated than breadth first or we may end up visiting `7` before `[5, 6]` (if your implementation avoids this, we should probably call it something like "topological" rather than "breadth-first").

Reply:
Hm, the point is I have multiple root specs, and need to handle the roots differently from the dependencies. Given that the second root may appear as a dependency of the first root, I don't want depth-first search. Breadth-first search initialized with all root specs guarantees that roots are found at depth 0 and that all connected nodes are visited once. I don't see the need for topological sort, for make targets all I need is the adjacency list.